staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1864,3 +1986,523 @@ func TestForgetWatcher(t *testing.T) {
+			opts:                      listOptions(false, pointer.Bool(false), "105"),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
+		},
+	}
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			backingStorage := &dummyStorage{}
+			cacher, _, err := newTestCacher(backingStorage)
+			require.NoError(t, err, "couldn't create cacher")
+			cacher.watchCache.resourceVersion = uint64(scenario.watchCacheResourceVersion)
+
+			parsedResourceVersion := 0
+			if len(scenario.opts.ResourceVersion) > 0 {
+				parsedResourceVersion, err = strconv.Atoi(scenario.opts.ResourceVersion)
+				require.NoError(t, err)
+			}
+
+			getBookMarkFn, err := cacher.getBookmarkAfterResourceVersionLockedFunc(uint64(parsedResourceVersion), uint64(scenario.requiredResourceVersion), scenario.opts)
+			require.NoError(t, err)
+			require.Equal(t, uint64(scenario.expectedBookmarkResourceVersion), getBookMarkFn(), "received unexpected ResourceVersion")
+		})
+	}
+}
+
+// TestGetStartResourceVersionForWatchLockedFunc test the following cases:
+//
+// +-----------------+---------------------+-----------------------+
+// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+// +=================+=====================+=======================+
+// | Unset           | true/false          | nil/true/false        |
+// | 0               | true/false          | nil/true/false        |
+// | 105             | true/false          | nil/true/false        |
+// +-----------------+---------------------+-----------------------+
+// where:
+// - false indicates the value of the param was set to "false" by a test case
+// - true  indicates the value of the param was set to "true" by a test case
+func TestGetStartResourceVersionForWatchLockedFunc(t *testing.T) {
+	listOptions := func(allowBookmarks bool, sendInitialEvents *bool, rv string) storage.ListOptions {
+		p := storage.Everything
+		p.AllowWatchBookmarks = allowBookmarks
+
+		opts := storage.ListOptions{}
+		opts.Predicate = p
+		opts.SendInitialEvents = sendInitialEvents
+		opts.ResourceVersion = rv
+		return opts
+	}
+
+	scenarios := []struct {
+		name                      string
+		opts                      storage.ListOptions
+		requiredResourceVersion   int
+		watchCacheResourceVersion int
+
+		expectedStartResourceVersion int
+	}{
+		// +-----------------+---------------------+-----------------------+
+		// | ResourceVersion | AllowWatchBookmarks | SendInitialEvents     |
+		// +=================+=====================+=======================+
+		// | Unset           | true/false          | nil/true/false        |
+		// +-----------------+---------------------+-----------------------+
+		{
+			name:                      "RV=unset, allowWatchBookmarks=true, sendInitialEvents=nil, expectedStartRV=0",
+			opts:                      listOptions(true, nil, ""),
+			requiredResourceVersion:   100,
+			watchCacheResourceVersion: 90,
```

Comment:
This also seems wrong - with RV unset, this should be storage version (so 100). 

For a bunch of other tests below too...

Reply:
I think this translates to `Get State and Start at Most Recent` semantics, i think that we require to return all events up to the RV. I think this our legacy case.

Reply:
I pointed you to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L517

With the change of "first wait for RV that is what you requested and only the serve the data", every single flow is:
- wait for RV
- serve "list" at that RV [if requested]
- server bookmark at that RV [if requested]
- continue with watch

Now - every case is exactly the same - I don't fully follow your comment.